src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java

```diff
@@ -0,0 +1,428 @@
+     * PerUnitMetricCollector should receive a specific record and distribute records on to the that collector
+     *
+     * There were will be 0 or 1 Distributors for each of the following MetriAcummulationLevels:
+     * ALL_READS, SAMPLE, LIBRARY, READ_GROUP
+     */
+    private abstract class Distributor implements Serializable {
+        private static final long serialVersionUID = 1L;
+
+        //A Map mapping the key for a specific record (as determined by getKey) to the appropriate collector
+        private final Map<String, UNIT_COLLECTOR> collectors;
+
+        //Given a SAMReadGroupRecord, return the key that identifies the collector for the corresponding SAMRecord
+        protected abstract String getKey(final SAMReadGroupRecord rg);
+
+        //Make a PerUnitMetricCollector for this given Distributor
+        protected abstract UNIT_COLLECTOR makeCollector(final SAMReadGroupRecord rg);
+
+        protected abstract UNIT_COLLECTOR makeUnknownCollector();
+
+        public Distributor(final List<SAMReadGroupRecord> rgRecs) {
+            collectors = new LinkedHashMap<>();
+            for(final SAMReadGroupRecord rg : rgRecs) {
+                final String key = getKey(rg);
+                if(!collectors.containsKey(key)) {
+                    collectors.put(key, makeCollector(rg));
+                }
+            }
+        }
+
+        /** Call acceptRecord(args) on the record collector identified by getKey */
+        public void acceptRecord(final ARGTYPE args, final SAMReadGroupRecord rg) {
+
+            String key = UNKNOWN;
+            if(rg != null) {
+                final String computedKey = getKey(rg);
+                if(computedKey != null) {
+                    key = computedKey;
+                }
+            }
+            UNIT_COLLECTOR collector = collectors.get(key);
+            if (collector == null) {
+                if (!UNKNOWN.equals(key)) {
+                    throw new GATKException("Could not find collector for " + key);
+                }
+                collector = makeUnknownCollector();
+                collectors.put(key, collector);
+            }
+            collector.acceptRecord(args);
+        }
+
+        /** Call finish on each PerUnitMetricCollector in this Aggregate Collector */
+        public void finish() {
+            for(final UNIT_COLLECTOR collector : collectors.values()) {
+                collector.finish();
+            }
+        }
+
+        /**
+         * Combine/merge the source distributor into this distributor by merging the collector maps.
+         * @param source source Distributor
+         * @param reMap map entry merge (reMap) function for combining distributor map entries. This
+         *              is usually the combineUnit method of MultiLevelCollectorReducible.
+         */
+        public void combine(Distributor source, BiFunction<UNIT_COLLECTOR, UNIT_COLLECTOR, UNIT_COLLECTOR> reMap) {
```

Comment:
I think the only place merge is used is when referring to arguments for the actual merge method on Map. Having said that, some of the comments in other places are not inconsistent. Will unify.